### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

In [2]:
# File to Load
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas DataFrames
school_data_df = pd.read_csv(school_data_to_load)
student_data_df = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
school_data_complete_df = pd.merge(student_data_df, school_data_df, how="left", on=["school_name", "school_name"])
school_data_complete_df.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Calculate the percentage of students who passed math **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [3]:
# Calculate the number of schools
schools_unique_df = school_data_complete_df["school_name"].unique()
total_schools_unique = school_data_complete_df["school_name"].nunique()
total_schools_unique


15

In [4]:
# Calculate the number of students
total_students = school_data_complete_df["student_name"].size
total_students

39170

In [5]:
# Calculate total budgets
school_budget_summary = school_data_df["budget"].sum()
school_budget_summary

24649428

In [6]:
# Calculate the average maths score for district summary
average_maths_score = school_data_complete_df["math_score"].mean()
average_maths_score

78.98537145774827

In [7]:
# Calculate the average reading score for district summary
average_reading_score = school_data_complete_df["reading_score"].mean()
average_reading_score

81.87784018381414

In [8]:
# Calculate the percentage of students in the district that have 70 or higher on their maths scores
passing_maths_total_df = school_data_complete_df.loc[school_data_complete_df["math_score"] >= 70,:]
passing_maths_total_students = passing_maths_total_df["student_name"].size
passing_maths_percentage = passing_maths_total_students/ total_students * 100
passing_maths_percentage


74.9808526933878

In [9]:
# Calculate the percentage of students in the district that have 70 or higher on their reading scores
passing_reading_total_df = school_data_complete_df.loc[school_data_complete_df["reading_score"] >= 70,:]
passing_reading_total_students = passing_reading_total_df["student_name"].size
passing_reading_percentage = passing_reading_total_students/ total_students * 100
passing_reading_percentage

85.80546336482001

In [10]:
# Calculate the percentage of students in the distrcict that have 70 or higher on BOTH their maths and reading scores
passing_both_total_df = school_data_complete_df.loc[(school_data_complete_df["math_score"] >= 70) & (school_data_complete_df["reading_score"] >= 70) ,:]
passing_both_total_students = passing_both_total_df["student_name"].size
passing_both_percentage = passing_both_total_students/ total_students * 100
passing_both_percentage

65.17232575950983

In [11]:
# Creating a new dataframe that contains the district summary data
district_summary_df = pd.DataFrame([[total_schools_unique,total_students,school_budget_summary,average_maths_score,average_reading_score,passing_maths_percentage,passing_reading_percentage,passing_both_percentage]], columns = ['Total Schools','Total Students','Total Budget','Average Maths Score','Average Reading Score','% Passing Maths','% Passing Reading','% Overall Passing'])
district_summary_df["Total Students"] = district_summary_df["Total Students"].astype(int).map("{:,}".format)
district_summary_df["Total Budget"] = district_summary_df["Total Budget"].astype(int).map("${:,.2f}".format)
district_summary_df

Total Schools Total Students    Total Budget  Average Maths Score  \
0             15         39,170  $24,649,428.00            78.985371   

   Average Reading Score  % Passing Maths  % Passing Reading  \
0               81.87784        74.980853          85.805463   

   % Overall Passing  
0          65.172326

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)
  
* Create a dataframe to hold the above results

In [12]:
# Calculate the average reading and maths scores for each school
group_by_schools_scores_df = school_data_complete_df.groupby('school_name')[["reading_score", "math_score"]].mean()
group_by_schools_scores_df

reading_score  math_score
school_name                                     
Bailey High School         81.033963   77.048432
Cabrera High School        83.975780   83.061895
Figueroa High School       81.158020   76.711767
Ford High School           80.746258   77.102592
Griffin High School        83.816757   83.351499
Hernandez High School      80.934412   77.289752
Holden High School         83.814988   83.803279
Huang High School          81.182722   76.629414
Johnson High School        80.966394   77.072464
Pena High School           84.044699   83.839917
Rodriguez High School      80.744686   76.842711
Shelton High School        83.725724   83.359455
Thomas High School         83.848930   83.418349
Wilson High School         83.989488   83.274201
Wright High School         83.955000   83.682222

In [13]:
school_data_df


School ID            school_name      type  size   budget
0           0      Huang High School  District  2917  1910635
1           1   Figueroa High School  District  2949  1884411
2           2    Shelton High School   Charter  1761  1056600
3           3  Hernandez High School  District  4635  3022020
4           4    Griffin High School   Charter  1468   917500
5           5     Wilson High School   Charter  2283  1319574
6           6    Cabrera High School   Charter  1858  1081356
7           7     Bailey High School  District  4976  3124928
8           8     Holden High School   Charter   427   248087
9           9       Pena High School   Charter   962   585858
10         10     Wright High School   Charter  1800  1049400
11         11  Rodriguez High School  District  3999  2547363
12         12    Johnson High School  District  4761  3094650
13         13       Ford High School  District  2739  1763916
14         14     Thomas High School   Charter  1635  1043130

In [14]:
# Merging student dataframe with data frames with scores average by school
group_by_school_merge_df = pd.merge(school_data_df,group_by_schools_scores_df,how="left", on=["school_name", "school_name"])
group_by_school_merge_df

School ID            school_name      type  size   budget  reading_score  \
0           0      Huang High School  District  2917  1910635      81.182722   
1           1   Figueroa High School  District  2949  1884411      81.158020   
2           2    Shelton High School   Charter  1761  1056600      83.725724   
3           3  Hernandez High School  District  4635  3022020      80.934412   
4           4    Griffin High School   Charter  1468   917500      83.816757   
5           5     Wilson High School   Charter  2283  1319574      83.989488   
6           6    Cabrera High School   Charter  1858  1081356      83.975780   
7           7     Bailey High School  District  4976  3124928      81.033963   
8           8     Holden High School   Charter   427   248087      83.814988   
9           9       Pena High School   Charter   962   585858      84.044699   
10         10     Wright High School   Charter  1800  1049400      83.955000   
11         11  Rodriguez High School  District  3999  2547363      80.744686   
12         12    Johnson High School  District  4761  3094650      80.966394   
13         13       Ford High School  District  2739  1763916      80.746258   
14         14     Thomas High School   Charter  1635  1043130      83.848930   

    math_score  
0    76.629414  
1    76.711767  
2    83.359455  
3    77.289752  
4    83.351499  
5    83.274201  
6    83.061895  
7    77.048432  
8    83.803279  
9    83.839917  
10   83.682222  
11   76.842711  
12   77.072464  
13   77.102592  
14   83.418349

In [15]:
# Dropping "School ID" column from the newly merged dataframe
group_by_schools_merge_df = group_by_school_merge_df.drop(columns='School ID')
group_by_schools_merge_df

school_name      type  size   budget  reading_score  math_score
0       Huang High School  District  2917  1910635      81.182722   76.629414
1    Figueroa High School  District  2949  1884411      81.158020   76.711767
2     Shelton High School   Charter  1761  1056600      83.725724   83.359455
3   Hernandez High School  District  4635  3022020      80.934412   77.289752
4     Griffin High School   Charter  1468   917500      83.816757   83.351499
5      Wilson High School   Charter  2283  1319574      83.989488   83.274201
6     Cabrera High School   Charter  1858  1081356      83.975780   83.061895
7      Bailey High School  District  4976  3124928      81.033963   77.048432
8      Holden High School   Charter   427   248087      83.814988   83.803279
9        Pena High School   Charter   962   585858      84.044699   83.839917
10     Wright High School   Charter  1800  1049400      83.955000   83.682222
11  Rodriguez High School  District  3999  2547363      80.744686   76.842711
12    Johnson High School  District  4761  3094650      80.966394   77.072464
13       Ford High School  District  2739  1763916      80.746258   77.102592
14     Thomas High School   Charter  1635  1043130      83.848930   83.418349

In [16]:
# Calculate the per student budget 
per_student_budget = (group_by_schools_merge_df['budget'] / group_by_schools_merge_df['size']).to_frame('Per Student Budget')
per_student_budget

Per Student Budget
0                655.0
1                639.0
2                600.0
3                652.0
4                625.0
5                578.0
6                582.0
7                628.0
8                581.0
9                609.0
10               583.0
11               637.0
12               650.0
13               644.0
14               638.0

In [17]:
# Extracting the necessary column from the per_student_budget dataframe
extracted_col = per_student_budget["Per Student Budget"]

In [18]:
# Joining the "Per Student Budget" column with the group by schools dataframe
group_by_schools_merge_df = group_by_schools_merge_df.join(extracted_col)
group_by_schools_merge_df

school_name      type  size   budget  reading_score  math_score  \
0       Huang High School  District  2917  1910635      81.182722   76.629414   
1    Figueroa High School  District  2949  1884411      81.158020   76.711767   
2     Shelton High School   Charter  1761  1056600      83.725724   83.359455   
3   Hernandez High School  District  4635  3022020      80.934412   77.289752   
4     Griffin High School   Charter  1468   917500      83.816757   83.351499   
5      Wilson High School   Charter  2283  1319574      83.989488   83.274201   
6     Cabrera High School   Charter  1858  1081356      83.975780   83.061895   
7      Bailey High School  District  4976  3124928      81.033963   77.048432   
8      Holden High School   Charter   427   248087      83.814988   83.803279   
9        Pena High School   Charter   962   585858      84.044699   83.839917   
10     Wright High School   Charter  1800  1049400      83.955000   83.682222   
11  Rodriguez High School  District  3999  2547363      80.744686   76.842711   
12    Johnson High School  District  4761  3094650      80.966394   77.072464   
13       Ford High School  District  2739  1763916      80.746258   77.102592   
14     Thomas High School   Charter  1635  1043130      83.848930   83.418349   

    Per Student Budget  
0                655.0  
1                639.0  
2                600.0  
3                652.0  
4                625.0  
5                578.0  
6                582.0  
7                628.0  
8                581.0  
9                609.0  
10               583.0  
11               637.0  
12               650.0  
13               644.0  
14               638.0

In [19]:
# Count the total number of students in each school
total_students_per_school = student_data_df.groupby("school_name").count()
total_students_per_school

Student ID  student_name  gender  grade  reading_score  \
school_name                                                                     
Bailey High School           4976          4976    4976   4976           4976   
Cabrera High School          1858          1858    1858   1858           1858   
Figueroa High School         2949          2949    2949   2949           2949   
Ford High School             2739          2739    2739   2739           2739   
Griffin High School          1468          1468    1468   1468           1468   
Hernandez High School        4635          4635    4635   4635           4635   
Holden High School            427           427     427    427            427   
Huang High School            2917          2917    2917   2917           2917   
Johnson High School          4761          4761    4761   4761           4761   
Pena High School              962           962     962    962            962   
Rodriguez High School        3999          3999    3999   3999           3999   
Shelton High School          1761          1761    1761   1761           1761   
Thomas High School           1635          1635    1635   1635           1635   
Wilson High School           2283          2283    2283   2283           2283   
Wright High School           1800          1800    1800   1800           1800   

                       math_score  
school_name                        
Bailey High School           4976  
Cabrera High School          1858  
Figueroa High School         2949  
Ford High School             2739  
Griffin High School          1468  
Hernandez High School        4635  
Holden High School            427  
Huang High School            2917  
Johnson High School          4761  
Pena High School              962  
Rodriguez High School        3999  
Shelton High School          1761  
Thomas High School           1635  
Wilson High School           2283  
Wright High School           1800

In [20]:
# Extracting a column with the total number of students per school
total_students_per_school = total_students_per_school["Student ID"]
total_students_per_school

school_name
Bailey High School       4976
Cabrera High School      1858
Figueroa High School     2949
Ford High School         2739
Griffin High School      1468
Hernandez High School    4635
Holden High School        427
Huang High School        2917
Johnson High School      4761
Pena High School          962
Rodriguez High School    3999
Shelton High School      1761
Thomas High School       1635
Wilson High School       2283
Wright High School       1800
Name: Student ID, dtype: int64

In [21]:
# Filtering the school data complete dataframe by only students who got 70 or higher on their maths scores 
school_data_complete_passing_maths = school_data_complete_df.loc[school_data_complete_df["math_score"] >= 70,:]
school_data_complete_passing_maths
# Count the number of students who got 70 or higher on their maths scores per school 
group_by_passing_maths = school_data_complete_passing_maths.groupby("school_name").count()
group_by_passing_maths
# Extracting a column with the number of students who got 70 or higher on their maths scores per school
group_by_passing_maths = group_by_passing_maths ["Student ID"] 
group_by_passing_maths
# Calculate the percentage of students who got 70 or higher on their maths scores per school
percent_passing_maths_df = group_by_passing_maths / total_students_per_school * 100
percent_passing_maths_df

school_name
Bailey High School       66.680064
Cabrera High School      94.133477
Figueroa High School     65.988471
Ford High School         68.309602
Griffin High School      93.392371
Hernandez High School    66.752967
Holden High School       92.505855
Huang High School        65.683922
Johnson High School      66.057551
Pena High School         94.594595
Rodriguez High School    66.366592
Shelton High School      93.867121
Thomas High School       93.272171
Wilson High School       93.867718
Wright High School       93.333333
Name: Student ID, dtype: float64

In [22]:
# Filtering the school data complete dataframe by only students who got 70 or higher on their reading scores 
school_data_complete_df_passing_reading = school_data_complete_df.loc[school_data_complete_df["reading_score"] >= 70,:]
school_data_complete_df_passing_reading
# Count the number of students who got 70 or higher on their reading scores per school 
group_by_passing_reading = school_data_complete_df_passing_reading.groupby("school_name").count()
group_by_passing_reading
# Extracting a column with the number of students who got 70 or higher on their reading scores per school
group_by_passing_reading = group_by_passing_reading["Student ID"]
group_by_passing_reading
# Calculate the percentage of students who got 70 or higher on their reading scores per school
percent_passing_reading_df = group_by_passing_reading / total_students_per_school * 100
percent_passing_reading_df

school_name
Bailey High School       81.933280
Cabrera High School      97.039828
Figueroa High School     80.739234
Ford High School         79.299014
Griffin High School      97.138965
Hernandez High School    80.862999
Holden High School       96.252927
Huang High School        81.316421
Johnson High School      81.222432
Pena High School         95.945946
Rodriguez High School    80.220055
Shelton High School      95.854628
Thomas High School       97.308869
Wilson High School       96.539641
Wright High School       96.611111
Name: Student ID, dtype: float64

In [23]:
# Filtering the school data complete dataframe by only students who got 70 or higher on both their maths and reading scores 
school_data_complete_df_passing_overall = school_data_complete_df.loc[(school_data_complete_df["reading_score"] >= 70) & (school_data_complete_df["math_score"]>= 70),:]
school_data_complete_df_passing_overall
# Count the number of students who got 70 or higher on both their maths and reading scores per school
group_by_passing_overall = school_data_complete_df_passing_overall.groupby("school_name").count()
group_by_passing_overall
# Extracting a column with the number of students who got 70 or higher on both their maths and reading scores per school
group_by_passing_overall = group_by_passing_overall["Student ID"]
group_by_passing_overall
# Calculate the percentage of students who got 70 or higher on both their maths and reading scores per school
percent_passing_overall_df = group_by_passing_overall / total_students_per_school * 100
percent_passing_overall_df

school_name
Bailey High School       54.642283
Cabrera High School      91.334769
Figueroa High School     53.204476
Ford High School         54.289887
Griffin High School      90.599455
Hernandez High School    53.527508
Holden High School       89.227166
Huang High School        53.513884
Johnson High School      53.539172
Pena High School         90.540541
Rodriguez High School    52.988247
Shelton High School      89.892107
Thomas High School       90.948012
Wilson High School       90.582567
Wright High School       90.333333
Name: Student ID, dtype: float64

In [24]:
# Creating a school summary dataframe by merging the group_by_schools dataframe with the percent passing maths dataframe
school_summary_df = pd.merge(group_by_schools_merge_df,percent_passing_maths_df, on = "school_name", how = "outer" )
# Renaming the column to its appropriate name
school_summary_df = school_summary_df.rename(columns = {"Student ID" : "% Passing Math"})
school_summary_df

school_name      type  size   budget  reading_score  math_score  \
0       Huang High School  District  2917  1910635      81.182722   76.629414   
1    Figueroa High School  District  2949  1884411      81.158020   76.711767   
2     Shelton High School   Charter  1761  1056600      83.725724   83.359455   
3   Hernandez High School  District  4635  3022020      80.934412   77.289752   
4     Griffin High School   Charter  1468   917500      83.816757   83.351499   
5      Wilson High School   Charter  2283  1319574      83.989488   83.274201   
6     Cabrera High School   Charter  1858  1081356      83.975780   83.061895   
7      Bailey High School  District  4976  3124928      81.033963   77.048432   
8      Holden High School   Charter   427   248087      83.814988   83.803279   
9        Pena High School   Charter   962   585858      84.044699   83.839917   
10     Wright High School   Charter  1800  1049400      83.955000   83.682222   
11  Rodriguez High School  District  3999  2547363      80.744686   76.842711   
12    Johnson High School  District  4761  3094650      80.966394   77.072464   
13       Ford High School  District  2739  1763916      80.746258   77.102592   
14     Thomas High School   Charter  1635  1043130      83.848930   83.418349   

    Per Student Budget  % Passing Math  
0                655.0       65.683922  
1                639.0       65.988471  
2                600.0       93.867121  
3                652.0       66.752967  
4                625.0       93.392371  
5                578.0       93.867718  
6                582.0       94.133477  
7                628.0       66.680064  
8                581.0       92.505855  
9                609.0       94.594595  
10               583.0       93.333333  
11               637.0       66.366592  
12               650.0       66.057551  
13               644.0       68.309602  
14               638.0       93.272171

In [25]:
# Merging the percentage of passing reading into the school summary dataframe
school_summary_df = pd.merge(school_summary_df,percent_passing_reading_df, on = "school_name", how = "outer" )
# Renaming the column to its appropriate name
school_summary_df = school_summary_df.rename(columns = {"Student ID" : "% Passing Reading"})
school_summary_df

school_name      type  size   budget  reading_score  math_score  \
0       Huang High School  District  2917  1910635      81.182722   76.629414   
1    Figueroa High School  District  2949  1884411      81.158020   76.711767   
2     Shelton High School   Charter  1761  1056600      83.725724   83.359455   
3   Hernandez High School  District  4635  3022020      80.934412   77.289752   
4     Griffin High School   Charter  1468   917500      83.816757   83.351499   
5      Wilson High School   Charter  2283  1319574      83.989488   83.274201   
6     Cabrera High School   Charter  1858  1081356      83.975780   83.061895   
7      Bailey High School  District  4976  3124928      81.033963   77.048432   
8      Holden High School   Charter   427   248087      83.814988   83.803279   
9        Pena High School   Charter   962   585858      84.044699   83.839917   
10     Wright High School   Charter  1800  1049400      83.955000   83.682222   
11  Rodriguez High School  District  3999  2547363      80.744686   76.842711   
12    Johnson High School  District  4761  3094650      80.966394   77.072464   
13       Ford High School  District  2739  1763916      80.746258   77.102592   
14     Thomas High School   Charter  1635  1043130      83.848930   83.418349   

    Per Student Budget  % Passing Math  % Passing Reading  
0                655.0       65.683922          81.316421  
1                639.0       65.988471          80.739234  
2                600.0       93.867121          95.854628  
3                652.0       66.752967          80.862999  
4                625.0       93.392371          97.138965  
5                578.0       93.867718          96.539641  
6                582.0       94.133477          97.039828  
7                628.0       66.680064          81.933280  
8                581.0       92.505855          96.252927  
9                609.0       94.594595          95.945946  
10               583.0       93.333333          96.611111  
11               637.0       66.366592          80.220055  
12               650.0       66.057551          81.222432  
13               644.0       68.309602          79.299014  
14               638.0       93.272171          97.308869

In [26]:
# Merging the percentage of passing reading and maths into the school summary dataframe
school_summary_df = pd.merge(school_summary_df,percent_passing_overall_df, on = "school_name", how = "outer" )
# Renaming the column to its appropriate name
school_summary_df = school_summary_df.rename(columns = {"Student ID" : "% Overall Passing"})

In [27]:
# Renaming more columns to their appropriate names
school_summary_df = school_summary_df.rename(columns = {"reading_score" : "Average Reading Score", "math_score" : "Average Math Score", "type" : "School Type", "size": "Total Students", "budget" : "Total School Budget"})

In [28]:
# Moving around column and row positions and setting the school names as the indices.
columns_titles = ["school_name","School Type","Total Students","Total School Budget", "Per Student Budget", "Average Math Score","Average Reading Score","% Passing Math","% Passing Reading","% Overall Passing"]
school_summary_df = school_summary_df.reindex(columns=columns_titles)
school_summary_df = school_summary_df.sort_values("school_name") 
school_summary_df

school_name School Type  Total Students  Total School Budget  \
7      Bailey High School    District            4976              3124928   
6     Cabrera High School     Charter            1858              1081356   
1    Figueroa High School    District            2949              1884411   
13       Ford High School    District            2739              1763916   
4     Griffin High School     Charter            1468               917500   
3   Hernandez High School    District            4635              3022020   
8      Holden High School     Charter             427               248087   
0       Huang High School    District            2917              1910635   
12    Johnson High School    District            4761              3094650   
9        Pena High School     Charter             962               585858   
11  Rodriguez High School    District            3999              2547363   
2     Shelton High School     Charter            1761              1056600   
14     Thomas High School     Charter            1635              1043130   
5      Wilson High School     Charter            2283              1319574   
10     Wright High School     Charter            1800              1049400   

    Per Student Budget  Average Math Score  Average Reading Score  \
7                628.0           77.048432              81.033963   
6                582.0           83.061895              83.975780   
1                639.0           76.711767              81.158020   
13               644.0           77.102592              80.746258   
4                625.0           83.351499              83.816757   
3                652.0           77.289752              80.934412   
8                581.0           83.803279              83.814988   
0                655.0           76.629414              81.182722   
12               650.0           77.072464              80.966394   
9                609.0           83.839917              84.044699   
11               637.0           76.842711              80.744686   
2                600.0           83.359455              83.725724   
14               638.0           83.418349              83.848930   
5                578.0           83.274201              83.989488   
10               583.0           83.682222              83.955000   

    % Passing Math  % Passing Reading  % Overall Passing  
7        66.680064          81.933280          54.642283  
6        94.133477          97.039828          91.334769  
1        65.988471          80.739234          53.204476  
13       68.309602          79.299014          54.289887  
4        93.392371          97.138965          90.599455  
3        66.752967          80.862999          53.527508  
8        92.505855          96.252927          89.227166  
0        65.683922          81.316421          53.513884  
12       66.057551          81.222432          53.539172  
9        94.594595          95.945946          90.540541  
11       66.366592          80.220055          52.988247  
2        93.867121          95.854628          89.892107  
14       93.272171          97.308869          90.948012  
5        93.867718          96.539641          90.582567  
10       93.333333          96.611111          90.333333

In [29]:
school_summary_df_final = school_summary_df.set_index("school_name")
school_summary_df_final

School Type  Total Students  Total School Budget  \
school_name                                                              
Bailey High School       District            4976              3124928   
Cabrera High School       Charter            1858              1081356   
Figueroa High School     District            2949              1884411   
Ford High School         District            2739              1763916   
Griffin High School       Charter            1468               917500   
Hernandez High School    District            4635              3022020   
Holden High School        Charter             427               248087   
Huang High School        District            2917              1910635   
Johnson High School      District            4761              3094650   
Pena High School          Charter             962               585858   
Rodriguez High School    District            3999              2547363   
Shelton High School       Charter            1761              1056600   
Thomas High School        Charter            1635              1043130   
Wilson High School        Charter            2283              1319574   
Wright High School        Charter            1800              1049400   

                       Per Student Budget  Average Math Score  \
school_name                                                     
Bailey High School                  628.0           77.048432   
Cabrera High School                 582.0           83.061895   
Figueroa High School                639.0           76.711767   
Ford High School                    644.0           77.102592   
Griffin High School                 625.0           83.351499   
Hernandez High School               652.0           77.289752   
Holden High School                  581.0           83.803279   
Huang High School                   655.0           76.629414   
Johnson High School                 650.0           77.072464   
Pena High School                    609.0           83.839917   
Rodriguez High School               637.0           76.842711   
Shelton High School                 600.0           83.359455   
Thomas High School                  638.0           83.418349   
Wilson High School                  578.0           83.274201   
Wright High School                  583.0           83.682222   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Bailey High School             81.933280          54.642283  
Cabrera High School            97.039828          91.334769  
Figueroa High School           80.739234          53.204476  
Ford High School               79.299014          54.289887  
Griffin High School            97.138965          90.599455  
Hernandez High School          80.862999          53.527508  
Holden High School             96.252927          89.227166  
Huang High

In [30]:
# Formatting
school_summary_df_final["Total School Budget"] = school_summary_df_final["Total School Budget"].astype(int).map("${:,.2f}".format)
school_summary_df_final["Per Student Budget"] = school_summary_df_final["Per Student Budget"].astype(int).map("${:,.2f}".format)
# Final version of the school summary dataframe
school_summary_df_final

School Type  Total Students Total School Budget  \
school_name                                                             
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Bailey High School               $628.00           77.048432   
Cabrera High School              $582.00           83.061895   
Figueroa High School             $639.00           76.711767   
Ford High School                 $644.00           77.102592   
Griffin High School              $625.00           83.351499   
Hernandez High School            $652.00           77.289752   
Holden High School               $581.00           83.803279   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Pena High School                 $609.00           83.839917   
Rodriguez High School            $637.00           76.842711   
Shelton High School              $600.00           83.359455   
Thomas High School               $638.00           83.418349   
Wilson High School               $578.00           83.274201   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Bailey High School             81.933280          54.642283  
Cabrera High School            97.039828          91.334769  
Figueroa High School           80.739234          53.204476  
Ford High School               79.299014          54.289887  
Griffin High School            97.138965          90.599455  
Hernandez High School          80.862999          53.527508  
Holden High School             96.252927          89.227166  
Huang High School              81.316421    

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [31]:
# Dataframe containing the top 5 performing schools by % Overall Passing
top_performing_schools = school_summary_df.sort_values(['% Overall Passing'],ascending = False).head(5)
top_performing_schools["Total School Budget"] = top_performing_schools["Total School Budget"] .astype(int).map("${:,.2f}".format)
top_performing_schools["Per Student Budget"] = top_performing_schools["Per Student Budget"] .astype(int).map("${:,.2f}".format)
top_performing_schools = top_performing_schools.set_index("school_name")
top_performing_schools

School Type  Total Students Total School Budget  \
school_name                                                           
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   
Pena High School        Charter             962         $585,858.00   

                    Per Student Budget  Average Math Score  \
school_name                                                  
Cabrera High School            $582.00           83.061895   
Thomas High School             $638.00           83.418349   
Griffin High School            $625.00           83.351499   
Wilson High School             $578.00           83.274201   
Pena High School               $609.00           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading  \
school_name                                                                     
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   
Pena High School                 84.044699       94.594595          95.945946   

                     % Overall Passing  
school_name                             
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [32]:
# Dataframe containing the bottom 5 performing schools by % Overall Passing
bottom_performing_schools = school_summary_df.sort_values(['% Overall Passing'],ascending = True).head(5)
bottom_performing_schools["Total School Budget"] = bottom_performing_schools["Total School Budget"] .astype(int).map("${:,.2f}".format)
bottom_performing_schools["Per Student Budget"] = bottom_performing_schools["Per Student Budget"] .astype(int).map("${:,.2f}".format)
bottom_performing_schools = bottom_performing_schools.set_index("school_name")
bottom_performing_schools

School Type  Total Students Total School Budget  \
school_name                                                             
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Hernandez High School    District            4635       $3,022,020.00   
Johnson High School      District            4761       $3,094,650.00   

                      Per Student Budget  Average Math Score  \
school_name                                                    
Rodriguez High School            $637.00           76.842711   
Figueroa High School             $639.00           76.711767   
Huang High School                $655.00           76.629414   
Hernandez High School            $652.00           77.289752   
Johnson High School              $650.00           77.072464   

                       Average Reading Score  % Passing Math  \
school_name                                                    
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Rodriguez High School          80.220055          52.988247  
Figueroa High School           80.739234          53.204476  
Huang High School              81.316421          53.513884  
Hernandez High School          80.862999          53.527508  
Johnson High School            81.222432          53.539172

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [33]:
student_data_df

Student ID       student_name gender grade         school_name  \
0               0       Paul Bradley      M   9th   Huang High School   
1               1       Victor Smith      M  12th   Huang High School   
2               2    Kevin Rodriguez      M  12th   Huang High School   
3               3  Dr. Richard Scott      M  12th   Huang High School   
4               4         Bonnie Ray      F   9th   Huang High School   
...           ...                ...    ...   ...                 ...   
39165       39165       Donna Howard      F  12th  Thomas High School   
39166       39166          Dawn Bell      F  10th  Thomas High School   
39167       39167     Rebecca Tanner      F   9th  Thomas High School   
39168       39168       Desiree Kidd      F  10th  Thomas High School   
39169       39169    Carolyn Jackson      F  11th  Thomas High School   

       reading_score  math_score  
0                 66          79  
1                 94          61  
2                 90          60  
3                 67          58  
4                 97          84  
...              ...         ...  
39165             99          90  
39166             95          70  
39167             73          84  
39168             99          90  
39169             95          75  

[39170 rows x 7 columns]

In [34]:
# Selecting the students in 9th grade only and finding their average maths and reading scores
group_by_9th_grade = student_data_df.loc[student_data_df["grade"] == "9th", :]
average_score_9th_grade = group_by_9th_grade.groupby('school_name')[["math_score","reading_score"]].mean()
# Renaming the columns to their appropriate names
average_score_9th_grade = average_score_9th_grade.rename(columns = {"math_score" : "9th_grade_maths_score", "reading_score" : "9th_grade_reading_score"})


In [35]:
# Putting the 9th graders' maths and reading scores columns into their own dataframe 
average_maths_score_9th_grade = average_score_9th_grade["9th_grade_maths_score"]
average_reading_score_9th_grade = average_score_9th_grade["9th_grade_reading_score"]

In [36]:
# Selecting the students in 10th grade only and finding their average maths and reading scores
group_by_10th_grade = student_data_df.loc[student_data_df["grade"] == "10th", :]
average_score_10th_grade = group_by_10th_grade.groupby('school_name')[["math_score","reading_score"]].mean()
# Renaming the columns to their appropriate names
average_score_10th_grade = average_score_10th_grade.rename(columns = {"math_score" : "10th_grade_maths_score", "reading_score" : "10th_grade_reading_score"})

In [37]:
# Putting the 10th graders' maths and reading scores columns into their own dataframe 
average_maths_score_10th_grade = average_score_10th_grade["10th_grade_maths_score"]
average_reading_score_10th_grade = average_score_10th_grade["10th_grade_reading_score"]

In [38]:
# Selecting the students in 11th grade only and finding their average maths and reading scores
group_by_11th_grade = student_data_df.loc[student_data_df["grade"] == "11th", :]
average_score_11th_grade = group_by_11th_grade.groupby('school_name')[["math_score","reading_score"]].mean()
# Renaming the columns to their appropriate names
average_score_11th_grade = average_score_11th_grade.rename(columns = {"math_score" : "11th_grade_maths_score", "reading_score" : "11th_grade_reading_score"})


In [39]:
# Putting the 11th graders' maths and reading scores columns into their own dataframe 
average_maths_score_11th_grade = average_score_11th_grade["11th_grade_maths_score"]
average_reading_score_11th_grade = average_score_11th_grade["11th_grade_reading_score"]

In [40]:
# Selecting the students in 12th grade only and finding their average maths and reading scores
group_by_12th_grade = student_data_df.loc[student_data_df["grade"] == "12th", :]
average_score_12th_grade = group_by_12th_grade.groupby('school_name')[["math_score","reading_score"]].mean()
# Renaming the columns to their appropriate names
average_score_12th_grade = average_score_12th_grade.rename(columns = {"math_score" : "12th_grade_maths_score", "reading_score" : "12th_grade_reading_score"})

In [41]:
# Putting the 12th graders' maths and reading scores columns into their own dataframe 
average_maths_score_12th_grade = average_score_12th_grade["12th_grade_maths_score"]
average_reading_score_12th_grade = average_score_12th_grade["12th_grade_reading_score"]

In [42]:
# Merging the 9th and 10th graders' maths scores in into their own dataframe
merging_maths_score_9th_and_10th = pd.merge(average_maths_score_9th_grade,average_maths_score_10th_grade, on = "school_name", how = "left")
# Merging the 11th and 12th graders' maths scores in into their own dataframe
merging_maths_score_11th_and_12th = pd.merge(average_maths_score_11th_grade,average_maths_score_12th_grade, on = "school_name", how = "left")
# Now we merge the two dataframes above into one dataframe
average_maths_score_by_grade = pd.merge(merging_maths_score_9th_and_10th,merging_maths_score_11th_and_12th, on = "school_name", how = "outer")

In [43]:
# Renaming the columns to their approriate names
average_maths_score_by_grade = average_maths_score_by_grade.rename(columns = {"9th_grade_maths_score" : "9th", "10th_grade_maths_score" : "10th", "11th_grade_maths_score" : "11th", "12th_grade_maths_score":"12th"})
# Final dataframe version
average_maths_score_by_grade

9th       10th       11th       12th
school_name                                                      
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [44]:
# Merging the 9th and 10th graders' reading scores in into their own dataframe
merging_reading_score_9th_and_10th = pd.merge(average_reading_score_9th_grade,average_reading_score_10th_grade, on = "school_name", how = "left")
# Merging the 11th and 12th graders' maths scores in into their own dataframe
merging_reading_score_11th_and_12th = pd.merge(average_reading_score_11th_grade,average_reading_score_12th_grade, on = "school_name", how = "left")
# Now we merge the two dataframes above into one dataframe
average_reading_score_by_grade = pd.merge(merging_reading_score_9th_and_10th,merging_reading_score_11th_and_12th, on = "school_name", how = "outer")

In [45]:
# Renaming the columns to their approriate names
average_reading_score_by_grade = average_reading_score_by_grade.rename(columns = {"9th_grade_reading_score" : "9th", "10th_grade_reading_score" : "10th", "11th_grade_reading_score" : "11th", "12th_grade_reading_score":"12th"})
# Final dataframe version
average_reading_score_by_grade

9th       10th       11th       12th
school_name                                                      
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [46]:
school_spending_df = school_summary_df[["Per Student Budget", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]]
school_spending_df

Per Student Budget  Average Math Score  Average Reading Score  \
7                628.0           77.048432              81.033963   
6                582.0           83.061895              83.975780   
1                639.0           76.711767              81.158020   
13               644.0           77.102592              80.746258   
4                625.0           83.351499              83.816757   
3                652.0           77.289752              80.934412   
8                581.0           83.803279              83.814988   
0                655.0           76.629414              81.182722   
12               650.0           77.072464              80.966394   
9                609.0           83.839917              84.044699   
11               637.0           76.842711              80.744686   
2                600.0           83.359455              83.725724   
14               638.0           83.418349              83.848930   
5                578.0           83.274201              83.989488   
10               583.0           83.682222              83.955000   

    % Passing Math  % Passing Reading  % Overall Passing  
7        66.680064          81.933280          54.642283  
6        94.133477          97.039828          91.334769  
1        65.988471          80.739234          53.204476  
13       68.309602          79.299014          54.289887  
4        93.392371          97.138965          90.599455  
3        66.752967          80.862999          53.527508  
8        92.505855          96.252927          89.227166  
0        65.683922          81.316421          53.513884  
12       66.057551          81.222432          53.539172  
9        94.594595          95.945946          90.540541  
11       66.366592          80.220055          52.988247  
2        93.867121          95.854628          89.892107  
14       93.272171          97.308869          90.948012  
5        93.867718          96.539641          90.582567  
10       93.333333          96.611111          90.333333

In [47]:
# Create bins for spending ranges per student
bins_spending = [0, 585, 630, 645, 680]
bins_spending_labels = ["<$585", "$585-630", "$630-645", "$645-680"]

In [48]:
# Putting budget per student intheir respective bins
school_spending_df["Spending Ranges (Per Student)"] = pd.cut(school_spending_df["Per Student Budget"], bins_spending, labels = bins_spending_labels)
school_spending_df

C:\Users\Nhan Tran\AppData\Local\Temp\ipykernel_25660\3040967700.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  school_spending_df["Spending Ranges (Per Student)"] = pd.cut(school_spending_df["Per Student Budget"], bins_spending, labels = bins_spending_labels)


Per Student Budget  Average Math Score  Average Reading Score  \
7                628.0           77.048432              81.033963   
6                582.0           83.061895              83.975780   
1                639.0           76.711767              81.158020   
13               644.0           77.102592              80.746258   
4                625.0           83.351499              83.816757   
3                652.0           77.289752              80.934412   
8                581.0           83.803279              83.814988   
0                655.0           76.629414              81.182722   
12               650.0           77.072464              80.966394   
9                609.0           83.839917              84.044699   
11               637.0           76.842711              80.744686   
2                600.0           83.359455              83.725724   
14               638.0           83.418349              83.848930   
5                578.0           83.274201              83.989488   
10               583.0           83.682222              83.955000   

    % Passing Math  % Passing Reading  % Overall Passing  \
7        66.680064          81.933280          54.642283   
6        94.133477          97.039828          91.334769   
1        65.988471          80.739234          53.204476   
13       68.309602          79.299014          54.289887   
4        93.392371          97.138965          90.599455   
3        66.752967          80.862999          53.527508   
8        92.505855          96.252927          89.227166   
0        65.683922          81.316421          53.513884   
12       66.057551          81.222432          53.539172   
9        94.594595          95.945946          90.540541   
11       66.366592          80.220055          52.988247   
2        93.867121          95.854628          89.892107   
14       93.272171          97.308869          90.948012   
5        93.867718          96.539641          90.582567   
10       93.333333          96.611111          90.333333   

   Spending Ranges (Per Student)  
7                       $585-630  
6                          <$585  
1                       $630-645  
13                      $630-645  
4                       $585-630  
3                       $645-680  
8                          <$585  
0                       $645-680  
12                      $645-680  
9                       $585-630  
11                      $630-645  
2                       $585-630  
14                      $630-645  
5                          <$585  
10                         <$585

In [49]:
# Grouping the data by the spending ranges per student
school_spending_final_df = school_spending_df.groupby("Spending Ranges (Per Student)").mean()

In [50]:
# Formatting the dataframe
school_spending_final_df = school_spending_final_df.round(2)
school_spending_final_final_df = school_spending_final_df[["Average Math Score","Average Reading Score","% Passing Math" , "% Passing Reading" , "% Overall Passing"]]
# Final version of the dataframe
school_spending_final_final_df

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                       83.46                  83.93   
$585-630                                    81.90                  83.16   
$630-645                                    78.52                  81.62   
$645-680                                    77.00                  81.03   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                                   93.46              96.61   
$585-630                                87.13              92.72   
$630-645                                73.48              84.39   
$645-680                                66.16              81.13   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                      90.37  
$585-630                                   81.42  
$630-645                                   62.86  
$645-680                                   53.53

## Scores by School Size

* Perform the same operations as above, based on school size.

In [51]:
school_size_df = school_summary_df[["Total Students", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]]
school_size_df

Total Students  Average Math Score  Average Reading Score  % Passing Math  \
7             4976           77.048432              81.033963       66.680064   
6             1858           83.061895              83.975780       94.133477   
1             2949           76.711767              81.158020       65.988471   
13            2739           77.102592              80.746258       68.309602   
4             1468           83.351499              83.816757       93.392371   
3             4635           77.289752              80.934412       66.752967   
8              427           83.803279              83.814988       92.505855   
0             2917           76.629414              81.182722       65.683922   
12            4761           77.072464              80.966394       66.057551   
9              962           83.839917              84.044699       94.594595   
11            3999           76.842711              80.744686       66.366592   
2             1761           83.359455              83.725724       93.867121   
14            1635           83.418349              83.848930       93.272171   
5             2283           83.274201              83.989488       93.867718   
10            1800           83.682222              83.955000       93.333333   

    % Passing Reading  % Overall Passing  
7           81.933280          54.642283  
6           97.039828          91.334769  
1           80.739234          53.204476  
13          79.299014          54.289887  
4           97.138965          90.599455  
3           80.862999          53.527508  
8           96.252927          89.227166  
0           81.316421          53.513884  
12          81.222432          53.539172  
9           95.945946          90.540541  
11          80.220055          52.988247  
2           95.854628          89.892107  
14          97.308869          90.948012  
5           96.539641          90.582567  
10          96.611111          90.333333

In [52]:
# Create bins for spending ranges per student
bins_size = [0, 1000, 2000, 5000]
bins_size_labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [53]:
# Binning the number of students 
school_size_df["School Size"] = pd.cut(school_size_df["Total Students"], bins_size, labels = bins_size_labels)

C:\Users\Nhan Tran\AppData\Local\Temp\ipykernel_25660\2855703445.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  school_size_df["School Size"] = pd.cut(school_size_df["Total Students"], bins_size, labels = bins_size_labels)


In [54]:
# Grouping the school size and calculating the means in their respective bins
school_size_df = school_size_df.groupby("School Size").mean()
# Final dataframe version
school_size_final_df = school_size_df[["Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]]
school_size_final_df

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)               96.099437          89.883853  
Medium (1000-2000)          96.790680          90.621535  
Large (2000-5000)           82.766634          58.286003

## Scores by School Type

* Perform the same operations as above, based on school type

In [55]:
school_summary_df
# Create a data frame that includes all the revelant columns and grouping them by "School Type" and displaying their means
school_type_df = school_summary_df[["School Type", "Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]]
school_type_df = school_type_df.groupby("School Type").mean()
school_type_df

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter              96.586489          90.432244  
District             80.799062          53.672208

# Report

Based on the dataframes I've created, there are a few data that stood out for me.  First, the onverall passing rate for charter schools are much higher than the district schools (90% vs 54%).  Moreover, based on overall passing rate, the top 5 performing schools b are all charter schools while the bottom 5 performing schools are all district schools.  This is very surprising because we have almost the same number of district and charter schools (7 and 8 respectively).  I would be very interested to explore the reasons behind these data. A surprising discovery that I find is the inverse relationship between spending ranges (per student) and overall passing rate, meaning the more money spent per student the worse the overall passing rate is. I would have expected it to be opposite. In conclusion, I can firmly say that charter schools perform significantly better than district schools.  There are limitations to this dataset of course.  I could have been provided with the number of teachers per school, teacher salary, the income per family of the neighbourhood the schools are located in, etc... 